File: src/cli/files.go
package cli

import (
	"fmt"
	"strings"
)

// FileOps wraps file operations
type FileOps struct {
	recursive       bool
	force           bool
	allowGlob       bool
	suppressErrors  bool
	append          bool
}

// NewFileOps creates a new FileOps command builder
func NewFileOps() *FileOps {
	return &FileOps{
		force: true,
	}
}

// Recursive sets recursive mode
func (f *FileOps) Recursive() *FileOps {
	f.recursive = true
	return f
}

// Force sets force mode
func (f *FileOps) Force(value bool) *FileOps {
	f.force = value
	return f
}

// AllowGlob allows glob expansion
func (f *FileOps) AllowGlob() *FileOps {
	f.allowGlob = true
	return f
}

// SuppressErrors suppresses errors
func (f *FileOps) SuppressErrors() *FileOps {
	f.suppressErrors = true
	return f
}

// Append sets append mode
func (f *FileOps) Append() *FileOps {
	f.append = true
	return f
}

// Write generates command that writes content to a file
func (f *FileOps) Write(path, content string) string {
	sanitized := strings.ReplaceAll(content, "\\", "\\\\")
	sanitized = escapeSingleQuotes(sanitized)
	redir := ">>"
	if !f.append {
		redir = ">"
	}
	return fmt.Sprintf("printf '%s' %s %s", sanitized, redir, quotePath(path))
}

// Chmod generates chmod command
func (f *FileOps) Chmod(path, mode string) string {
	return fmt.Sprintf("chmod %s %s", mode, quotePath(path))
}

// Mkdir generates mkdir command
func (f *FileOps) Mkdir(path string, parents bool) string {
	flag := ""
	if parents {
		flag = "-p "
	}
	return fmt.Sprintf("mkdir %s%s", flag, quotePath(path))
}

// Chown generates chown command
func (f *FileOps) Chown(path, owner string, group *string) string {
	ownerSpec := owner
	if group != nil {
		ownerSpec = fmt.Sprintf("%s:%s", owner, *group)
	}
	return fmt.Sprintf("chown %s %s", ownerSpec, quotePath(path))
}

// Remove generates rm command
func (f *FileOps) Remove(path string) string {
	flags := ""
	if f.recursive {
		flags += "r"
	}
	if f.force {
		flags += "f"
	}
	flagPart := ""
	if flags != "" {
		flagPart = fmt.Sprintf("-%s ", flags)
	}
	quotedPath := path
	if !f.allowGlob {
		quotedPath = quotePath(path)
	}
	return fmt.Sprintf("rm %s%s", flagPart, quotedPath)
}

// Truncate generates truncate command
func (f *FileOps) Truncate(path string) string {
	cmd := fmt.Sprintf("truncate -s 0 %s", quotePath(path))
	return cmd
}

// Symlink generates symlink command
func (f *FileOps) Symlink(target, linkPath string) string {
	return fmt.Sprintf("ln -s %s %s", quotePath(target), quotePath(linkPath))
}

// FindDelete generates find delete command
func (f *FileOps) FindDelete(directory, pattern, fileType string) string {
	patternEscaped := escapeSingleQuotes(pattern)
	cmd := fmt.Sprintf("find %s -type %s -name '%s' -delete", quotePath(directory), fileType, patternEscaped)
	return cmd
}

// Exists generates command to check if file exists
func (f *FileOps) Exists(path string) string {
	cmd := fmt.Sprintf("test -f %s && echo exists || echo not_found", quotePath(path))
	return cmd
}

func quotePath(path string) string {
	// Simple quoting - in production use proper shell escaping
	if strings.Contains(path, " ") || strings.Contains(path, "$") {
		return fmt.Sprintf("'%s'", strings.ReplaceAll(path, "'", "'\"'\"'"))
	}
	return path
}

func escapeSingleQuotes(value string) string {
	return strings.ReplaceAll(value, "'", "'\"'\"'")
}

File: src/CLI/Files.cs
using System;

namespace Enva.CLI;

public static class Files
{
    public static FileOps NewFileOps()
    {
        return FileOps.NewFileOps();
    }
}

public class FileOps
{
    private string? path;
    private string? content;
    private string? mode;
    private bool parents;
    private bool append;
    private string operation = "";
    private bool recursive;
    private bool force = true;
    private bool allowGlob;

    public static FileOps NewFileOps()
    {
        return new FileOps();
    }

    public FileOps Recursive()
    {
        recursive = true;
        return this;
    }

    public FileOps Force(bool value)
    {
        force = value;
        return this;
    }

    public FileOps AllowGlob()
    {
        allowGlob = true;
        return this;
    }

    public FileOps SuppressErrors()
    {
        return this;
    }

    public FileOps Append()
    {
        append = true;
        return this;
    }

    public FileOps Mkdir(string dirPath, bool createParents = false)
    {
        path = dirPath;
        parents = createParents;
        operation = "mkdir";
        return this;
    }

    public FileOps Write(string filePath, string fileContent)
    {
        path = filePath;
        content = fileContent;
        append = false;
        operation = "write";
        return this;
    }

    public string Chmod(string filePath, string permissions)
    {
        return $"chmod {permissions} {Quote(filePath)}";
    }

    public string Remove(string filePath)
    {
        string flags = "";
        if (recursive)
        {
            flags += "r";
        }
        if (force)
        {
            flags += "f";
        }
        string flagPart = flags != "" ? $"-{flags} " : "";
        string quotedPath = allowGlob ? filePath : Quote(filePath);
        return $"rm {flagPart}{quotedPath}";
    }

    public string Truncate(string filePath)
    {
        return $"truncate -s 0 {Quote(filePath)}";
    }

    public string Symlink(string target, string linkPath)
    {
        return $"ln -s {Quote(target)} {Quote(linkPath)}";
    }

    public string FindDelete(string directory, string pattern, string fileType)
    {
        string patternEscaped = EscapeSingleQuotes(pattern);
        return $"find {Quote(directory)} -type {fileType} -name '{patternEscaped}' -delete";
    }

    public string ToCommand()
    {
        if (operation == "mkdir")
        {
            string flag = parents ? "-p " : "";
            return $"mkdir {flag}{Quote(path!)}";
        }
        if (operation == "write")
        {
            if (string.IsNullOrEmpty(path) || string.IsNullOrEmpty(content))
            {
                throw new InvalidOperationException("Path and content must be set for write operation");
            }
            string sanitized = content.Replace("\\", "\\\\");
            sanitized = EscapeSingleQuotes(sanitized);
            string redir = append ? ">>" : ">";
            return $"printf '{sanitized}' {redir} {Quote(path)}";
        }
        throw new InvalidOperationException("Invalid FileOps operation");
    }

    private string Quote(string s)
    {
        if (s.Contains(" ") || s.Contains("$") || s.Contains("'"))
        {
            return $"'{s.Replace("'", "'\"'\"'")}'";
        }
        return s;
    }

    private string EscapeSingleQuotes(string value)
    {
        return value.Replace("'", "'\"'\"'");
    }
}
